Don't Unnecessarily Invalidate Unreferenced Generic Parameters

philipturner opened the pull request:
<!-- What's in this pull request? -->
Currently, a `Decl` is set to invalid at the end of `TypeChecker::checkReferencedGenericParams`. That should not happen, and finally manifested in a crash discovered while testing AutoDiff:

```swift
import _Differentiation

struct Q { }

// Should issue a diagnostic, but crashes instead
@derivative(of: remainder(_:_:))
func _vjpRemainder<T: FloatingPoint>(_ x: Q, _ y: Q) -> (
  value: Q, pullback: (Q) -> (Q, Q)
) {
  fatalError()
}
```

As a result of this fix, error diagnostics for incorrect generic functions slightly change:

```swift
class SomeClassWithInvalidMethod {
  // NO LONGER EXISTS: expected-error {{generic parameter 'T' is not used in function signature}}
  // NOW EXISTS: expected-note {{in call to function 'method()'}}
  func method<T>() { 
    // NOW EXISTS: expected-error@-1 {{generic parameter 'T' is not used in function signature}}
    self.method()
    // NOW EXISTS: expected-error@-1 {{generic parameter 'T' could not be inferred}}
  }
}
```

And, this affects a battle between CSDiag and CSSolver:

```swift
// Fix for CSDiag vs CSSolver disagreement on what constitutes a
// valid overload.

// NO LONGER EXISTS: expected-note {{'overloadedMethod(n:)' declared here}}
func overloadedMethod(n: Int) {}

// NO LONGER EXISTS: expected-note {{in call to function 'overloadedMethod()'}}
// NOW EXISTS: {{generic parameter 'T' is not used in function signature}}
func overloadedMethod<T>() {}

// NO LONGER EXISTS: expected-error@-1 {{missing argument for parameter 'n' in call}}
// NOW EXISTS: expected-error@-1 {{generic parameter 'T' could not be inferred}}
overloadedMethod()
```

philipturner committed:
Attempt to fix a bug

rxwei reviewed (changes requested):
The diagnostic wasn't _missing_, it's a compiler bug. We should try to fix the bug instead.

philipturner commented:
Should a function be passing in an error in the first place? Also, please tell me if my pinging has been bothersome because I have no way of knowing right now.

AnthonyLatsis commented on lib/Sema/TypeCheckAttr.cpp:
```diff
@@ -4749,11 +4749,24 @@ static bool typeCheckDerivativeAttr(ASTContext &Ctx, Decl *D,
 
   auto *derivativeInterfaceType =
-      derivative->getInterfaceType()->castTo<AnyFunctionType>();
+      derivative->getInterfaceType()->getAs<AnyFunctionType>();
+  if (!derivativeInterfaceType) {
+    // Crashes if the instance is not an ErrorType.
+    if (derivative->getInterfaceType()->is<ErrorType>()) {
+      diags.diagnose(attr->getLocation(),
+                     diag::derivative_attr_unknown_error);
+      return true;
+    } else {
+      // This should never happen, but leaving a runtime diagnostic incase it
+      // does because little is known about this failure.
+      llvm::report_fatal_error("Unknown failure in typeCheckDerivativeAttr.");
+    }
+  }
```
If we encounter an `ErrorType` here, it means we already produced an appropriate diagnostic elsewhere. I think we should simply bail out if there's error in the interface type. Also, we should not be expecting `getInterfaceType()` to return a null type.

philipturner replied:
Calling `Decl->isInvalid()` returns false, so guarding against that does nothing for bailing out early. The only way to do so is to check that the pointer is null, then exit early. Since this behavior deserves to be caught, I'm issuing a second diagnostic.

philipturner replied:
`getInterfaceType()` is not what's returning a null. It's `getAs<...>` that's returning the null. In fact, I call a function on the result of `getInterfaceType()` in that conditional block.

AnthonyLatsis replied:
Are you sure? `Decl->isInvalid()` on a `FuncDecl` is equivalent to `getInterfaceType()->hasError()`. It *should* have an error with the code in your example.

philipturner replied:
So just check `hasError()`, silently bail if true, then cast to `AnyFunctionType` as was previously done?

AnthonyLatsis replied:
Sounds good. You can avoid defining an additional variable by checking the declaration instead (using `Decl->isInvalid()`).

philipturner replied:
I just said that Decl->isInvalid doesn't do anything useful. It always returns true here.

AnthonyLatsis replied:
> It always returns true here

That is exactly what you need. The declaration is invalid when its interface type contains an `ErrorType`.

philipturner replied:
I edited my response.

AnthonyLatsis replied:
Can you show me the exact code you're running?

philipturner replied:
The behavior just changed. It now does return true. Maybe I misremembered my experience debugging it.

AnthonyLatsis reviewed (commented):
Could you add a test as well?

philipturner commented:
I don't have much experience with detecting certain diagnostics in tests, but I could copy and paste some source code for you to refine.

AnthonyLatsis commented:
> I don't have much experience with detecting diagnostics in tests, but I could copy and paste some source code for you to refine.

You can add a test to `test/AutoDiff/Sema/derivative_attr_type_checking.swift` and run `lit` on it (see the [guide](https://github.com/apple/swift/blob/main/docs/HowToGuides/GettingStarted.md#running-tests) on how to run `lit` selectively). It will show you the missing diagnostics, and you can add them using the `expected-error` directive by following other examples in that file.

philipturner commented:
How to I incrementally build the compiler without rebuilding the standard library? I tried what it says on the README but it doesn't actually change anything. For example, I change the Swift version test, incrementally build, and it doesn't register a change. This is why it's @AnthonyLatsis this is why it's taking me several minutes to answer your question.

AnthonyLatsis commented:
> How to I incrementally build the compiler without rebuilding the standard library?

You use `ninja swift-frontend` to incrementally build the frontend (should be run in the `build/swift-macosx-x86_64` directory).

philipturner commented:
I've been running this command and it still keeps rebuilding Standard Library modules like SwiftOnoneSupport and _MatchingEngine:
```
ninja -C ../build/Ninja-RelWithDebInfoAssert/swift-macosx-$(uname -m)
```

philipturner commented:
I did this and something was different! It compiled in seconds!
```
cd ../build/Ninja-RelWithDebInfoAssert/swift-macosx-$(uname -m) && ninja swift-frontend
```

philipturner commented:
I did this. It compiled in seconds, then acted like nothing happened. For example, I removed a fatal error, yet the non-removed fatal error still fired.
```
cd ../build/Ninja-RelWithDebInfoAssert/swift-macosx-$(uname -m) && ninja swift-frontend
```

AnthonyLatsis commented:
What exactly are you doing to *run* the compiler? `swift-frontend` is the right target if you want to build just the compiler.

philipturner commented:
```
../build/Ninja-RelWithDebInfoAssert/swift-macosx-arm64/bin/swift -Xfrontend -requirement-machine-protocol-signatures=on -Xfrontend -requirement-machine-inferred-signatures=on -Xfrontend -requirement-machine-abstract-signatures=on -Xfrontend -enable-experimental-forward-mode-differentiation ../hello_world_test/file.swift
```

AnthonyLatsis commented:
If it's `build/swift-macosx-x86_64/bin/swift-frontend` or `build/swift-macosx-x86_64/bin/swiftc` you're running after building, and the output is still not reflecting your edits, then I'm confused.

philipturner commented:
And passing in "swift" to ninja gives an error.

AnthonyLatsis commented:
`swift-frontend` builds both the frontend and `swiftc`. There is no `swift` target. Does it not print anything if you add something like `llvm::errs()<<"Hello World\n";` to the code path you're testing?

philipturner commented:
It says it's compiling, and I can tell because it shows errors if I take a semicolon off of something. The problem might be that my command starts with this:
```
../build/Ninja-RelWithDebInfoAssert/swift-macosx-arm64/bin/swift
```
I'm not using swift-frontend or swiftc.

philipturner committed:
Fix some stuff

AnthonyLatsis commented:
It should be fine, `bin/swift` is an alias for `bin/swift-frontend`.

philipturner commented:
It actually isn't a direct alias.
```
(base) philipturner@philipsm1maxmbp swift % ../build/Ninja-RelWithDebInfoAssert/swift-macosx-arm64/bin/swift-frontend -Xfrontend -requirement-machine-protocol-signatures=on -Xfrontend -requirement-machine-inferred-signatures=on -Xfrontend -requirement-machine-abstract-signatures=on -Xfrontend -enable-experimental-forward-mode-differentiation ../hello_world_test/file.swift
<unknown>:0: error: unknown argument: '-Xfrontend'
<unknown>:0: error: unknown argument: '-Xfrontend'
<unknown>:0: error: unknown argument: '-Xfrontend'
<unknown>:0: error: unknown argument: '-Xfrontend'
(base) philipturner@philipsm1maxmbp swift % ../build/Ninja-RelWithDebInfoAssert/swift-macosx-arm64/bin/swift-frontend -requirement-machine-protocol-signatures=on -requirement-machine-inferred-signatures=on -requirement-machine-abstract-signatures=on -enable-experimental-forward-mode-differentiation ../hello_world_test/file.swift 
<unknown>:0: error: no frontend action was selected
```

AnthonyLatsis commented on lib/Sema/TypeCheckAttr.cpp:
```diff
@@ -4742,6 +4742,10 @@ void AttributeChecker::visitDifferentiableAttr(DifferentiableAttr *attr) {
+  if (D->isInvalid()) {
+    return true;
+  }
+  
```
We would want to check whether differentiable programming is enabled below before doing this.

philipturner commented:
I grabbed the hard-coded program arguments for swift-frontend (pasted below). I messed with the C++ code, but nothing changed:
```
/Users/philipturner/Documents/Swift-Compiler/swift-project/build/Ninja-RelWithDebInfoAssert/swift-macosx-arm64/bin/swift-frontend -frontend -interpret ../hello_world_test/file.swift -Xllvm -aarch64-use-tbi -enable-objc-interop -sdk /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX12.1.sdk -color-diagnostics -new-driver-path /Users/philipturner/Documents/Swift-Compiler/swift-project/build/Ninja-RelWithDebInfoAssert/swift-macosx-arm64/bin/swift-driver -requirement-machine-protocol-signatures=on -requirement-machine-inferred-signatures=on -requirement-machine-abstract-signatures=on -enable-experimental-forward-mode-differentiation -resource-dir /Users/philipturner/Documents/Swift-Compiler/swift-project/build/Ninja-RelWithDebInfoAssert/swift-macosx-arm64/lib/swift -module-name file -target-sdk-version 12.1
```

philipturner committed:
Relocate early return

philipturner commented:
@AnthonyLatsis could you try incrementally rebuilding the compiler yourself and attaching the build log?

AnthonyLatsis commented:
I can reproduce the crash if I run `ninja swift-frontend && bin/swift-frontend -typecheck path_to_file` with my build, and fix it by applying your changes and running the same commands. It also works if I use `swiftc` or `swift`.

philipturner commented:
This happens:
```
(base) philipturner@philipsm1maxmbp swift-macosx-arm64 % ninja swift-frontend && bin/swift-frontend -typecheck ../../../hello_world_test/file.swift
ninja: no work to do.
<unknown>:0: error: cannot load underlying module for 'Darwin'
<unknown>:0: note: did you forget to set an SDK using -sdk or SDKROOT?
<unknown>:0: note: use "xcrun swiftc" to select the default macOS SDK installed with Xcode
```

philipturner commented:
> I can reproduce the crash

I'm not talking about reproducing the crash. I'm talking about this:
1) You implement my changes that fix the crash
2) You fire up the Swift compiler and it's fine
3) You catastrophically destroy the C++ code, while letting it technically compile
4) You fire up the Swift compiler and `it's fine`!

AnthonyLatsis commented:
> This happens...

This is expected, my commands are enough to reproduce the crash and fix it, but you need to specify an SDK to properly compile AutoDiff code. `bin/swift` seems to be using the default SDK that ships with Xcode.

> I'm not talking about reproducing the crash. I'm talking about this...

What are the edits to the C++ code, and what is the unexpected output?

philipturner commented:
Go to the function we're debugging and put this at the start of it:
```
llvm::report_fatal_error("HELLO WORLD");
```
The expected error is that at the top of the crash log, you see "HELLO WORLD" somewhere. Instead, you don't.

AnthonyLatsis commented:
`ninja swift-frontend && bin/swift /Users/mac/Desktop/test.swift`:

```
/Users/mac/Desktop/test.swift:48:20: error: generic parameter 'T' is not used in function signature
func _vjpRemainder<T: FloatingPoint>(_ x: Q, _ y: Q) -> (
                   ^
<unknown>:0: error: fatal error encountered during compilation; please submit a bug report (https://swift.org/contributing/#reporting-bugs) and include the project
<unknown>:0: note: HELLO WORLD
Stack dump:
0.	Program arguments: /Users/mac/Desktop/swift-project/build/Ninja-RelWithDebInfoAssert/swift-macosx-x86_64/bin/swift-frontend -frontend -interpret /Users/mac/Desktop/test.swift -enable-objc-interop -sdk /Applications/Xcode-beta.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX12.0.sdk -color-diagnostics -target-sdk-version 12.0 -module-name test
1.	Swift version 5.7-dev (LLVM 51a24b2a8ac6130, Swift 5b350be3522123b)
2.	Compiling with the current language version
3.	While evaluating request TypeCheckSourceFileRequest(source_file "/Users/mac/Desktop/test.swift")
4.	While type-checking '_vjpRemainder(_:_:)' (at /Users/mac/Desktop/test.swift:48:1)
Stack dump without symbol names (ensure you have llvm-symbolizer in your PATH or set the environment var `LLVM_SYMBOLIZER_PATH` to point to it):
0  swift-frontend           0x000000010fd31f97 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) + 39
1  swift-frontend           0x000000010fd310c5 llvm::sys::RunSignalHandlers() + 85
2  swift-frontend           0x000000010fd325d0 SignalHandler(int) + 288
3  libsystem_platform.dylib 0x00007ff81eae5e2d _sigtramp + 29
4  libsystem_platform.dylib 000000000000000000 _sigtramp + 18446603370066190832
5  libsystem_c.dylib        0x00007ff81ea1cd10 abort + 123
6  swift-frontend           0x000000010a9c6948 swift::performFrontend(llvm::ArrayRef<char const*>, char const*, void*, swift::FrontendObserver*)::$_3::__invoke(void*, char const*, bool) + 552
7  swift-frontend           0x000000010fc7933d llvm::report_fatal_error(llvm::Twine const&, bool) + 365
8  swift-frontend           0x000000010fc791cb llvm::report_fatal_error(char const*, bool) + 43
9  swift-frontend           0x000000010b777ba5 typeCheckDerivativeAttr(swift::ASTContext&, swift::Decl*, swift::DerivativeAttr*) + 21
10 swift-frontend           0x000000010b766a4f swift::ASTVisitor<(anonymous namespace)::AttributeChecker, void, void, void, void, void, void>::visit(swift::DeclAttribute*) + 33519
11 swift-frontend           0x000000010b75dfb4 swift::TypeChecker::checkDeclAttributes(swift::Decl*) + 180
12 swift-frontend           0x000000010b7f0fa7 (anonymous namespace)::DeclChecker::visitFuncDecl(swift::FuncDecl*) + 135
13 swift-frontend           0x000000010b7e7c16 (anonymous namespace)::DeclChecker::visit(swift::Decl*) + 230
14 swift-frontend           0x000000010b7e7b14 swift::TypeChecker::typeCheckDecl(swift::Decl*, bool) + 116
15 swift-frontend           0x000000010b8aa53a swift::TypeCheckSourceFileRequest::evaluate(swift::Evaluator&, swift::SourceFile*) const + 186
16 swift-frontend           0x000000010b8ac370 llvm::Expected<swift::TypeCheckSourceFileRequest::OutputType> swift::Evaluator::getResultUncached<swift::TypeCheckSourceFileRequest>(swift::TypeCheckSourceFileRequest const&) + 416
17 swift-frontend           0x000000010b8ac124 llvm::Expected<swift::TypeCheckSourceFileRequest::OutputType> swift::Evaluator::getResultCached<swift::TypeCheckSourceFileRequest, (void*)0>(swift::TypeCheckSourceFileRequest const&) + 100
18 swift-frontend           0x000000010b8aa38f swift::TypeCheckSourceFileRequest::OutputType swift::evaluateOrDefault<swift::TypeCheckSourceFileRequest>(swift::Evaluator&, swift::TypeCheckSourceFileRequest, swift::TypeCheckSourceFileRequest::OutputType) + 31
19 swift-frontend           0x000000010aa2e0ec bool llvm::function_ref<bool (swift::SourceFile&)>::callback_fn<swift::CompilerInstance::performSema()::$_6>(long, swift::SourceFile&) + 12
20 swift-frontend           0x000000010aa26eee swift::CompilerInstance::forEachFileToTypeCheck(llvm::function_ref<bool (swift::SourceFile&)>) + 174
21 swift-frontend           0x000000010aa26d8a swift::CompilerInstance::performSema() + 74
22 swift-frontend           0x000000010a9c0d56 swift::performFrontend(llvm::ArrayRef<char const*>, char const*, void*, swift::FrontendObserver*) + 3510
23 swift-frontend           0x000000010a93c5f7 swift::mainEntry(int, char const**) + 503
24 dyld                     0x000000012d0c84fe start + 462
```

P.S. I tried your command as well just in case, the output is the same.

philipturner commented:
> bin/swift seems to be using the default SDK that ships with Xcode.

I just changed `bin/swift` to `bin/swift-frontend` and prepended it with `xcrun`. The command runs now, but it still doesn't crash when I modified the C++ code to crash:
```
ninja swift-frontend && xcrun bin/swift-frontend -emit-assembly ../../../hello_world_test/file.swift 
```

AnthonyLatsis commented:
Can you try something else like `llvm::errs()<<"HELLO WORLD\n";` or `llvm_unreachable("HELLO WORLD");`, or even just `assert(false && "HELLO WORLD");`?

philipturner commented:
Still no change. Were you able to get a change in behavior?

AnthonyLatsis commented:
>  Were you able to get a change in behavior?

No problem with the fatal error or anything else on my end. Did you notice it in the logs I attached?

philipturner commented:
Same commands as you. As this point, I'm giving up. I'd rather have your help with invesigating a new compiler crash. I just have to make modifications that don't corrupt the standard library, then cancel compilation when I know it's compiling the standard library.
```
(base) philipturner@philipsm1maxmbp swift-macosx-arm64 % ninja swift-frontend && bin/swift ../../../hello_world_test/file.swift
[1/3] Building CXX object lib/Sema/CMakeFiles/swiftSema.dir/TypeCheckAttr.cpp.o
/Users/philipturner/Documents/Swift-Compiler/swift-project/swift/lib/Sema/TypeCheckAttr.cpp:4748:17: warning: code will never be executed [-Wunreachable-code]
  auto &diags = Ctx.Diags;
                ^~~
1 warning generated.
[3/3] Linking CXX executable bootstrapping0/bin/swift-frontend
../../../hello_world_test/file.swift:7:20: error: generic parameter 'T' is not used in function signature
func _vjpRemainder<T: FloatingPoint>(_ x: Q, _ y: Q) -> (
                   ^
Assertion failed: (isa<X>(Val) && "cast<Ty>() argument of incompatible type!"), function cast, file Casting.h, line 269.
Stack dump:
0.	Program arguments: /Users/philipturner/Documents/Swift-Compiler/swift-project/build/Ninja-RelWithDebInfoAssert/swift-macosx-arm64/bin/swift-frontend -frontend -interpret ../../../hello_world_test/file.swift -Xllvm -aarch64-use-tbi -enable-objc-interop -sdk /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX12.1.sdk -color-diagnostics -new-driver-path /Users/philipturner/Documents/Swift-Compiler/swift-project/build/Ninja-RelWithDebInfoAssert/swift-macosx-arm64/bin/swift-driver -resource-dir /Users/philipturner/Documents/Swift-Compiler/swift-project/build/Ninja-RelWithDebInfoAssert/swift-macosx-arm64/lib/swift -module-name file -target-sdk-version 12.1
1.	Swift version 5.7-dev (LLVM 59972a8d54ac3bf, Swift edf98aa17cb6668)
2.	Compiling with the current language version
3.	While evaluating request TypeCheckSourceFileRequest(source_file "../../../hello_world_test/file.swift")
4.	While type-checking '_vjpRemainder(_:_:)' (at ../../../hello_world_test/file.swift:7:1)
Stack dump without symbol names (ensure you have llvm-symbolizer in your PATH or set the environment var `LLVM_SYMBOLIZER_PATH` to point to it):
0  swift-frontend           0x0000000106c1ffc0 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) + 56
1  swift-frontend           0x0000000106c1f0f0 llvm::sys::RunSignalHandlers() + 128
2  swift-frontend           0x0000000106c20624 SignalHandler(int) + 304
3  libsystem_platform.dylib 0x00000001bb5304e4 _sigtramp + 56
4  libsystem_pthread.dylib  0x00000001bb518eb0 pthread_kill + 288
5  libsystem_c.dylib        0x00000001bb456314 abort + 164
6  libsystem_c.dylib        0x00000001bb45572c err + 0
7  swift-frontend           0x000000010706d644 swift::ASTVisitor<(anonymous namespace)::AttributeChecker, void, void, void, void, void, void>::visit(swift::DeclAttribute*) (.cold.278) + 0
8  swift-frontend           0x0000000102f23170 swift::ASTVisitor<(anonymous namespace)::AttributeChecker, void, void, void, void, void, void>::visit(swift::DeclAttribute*) + 36136
9  swift-frontend           0x0000000102f19e10 swift::TypeChecker::checkDeclAttributes(swift::Decl*) + 184
10 swift-frontend           0x0000000102fa4b9c (anonymous namespace)::DeclChecker::visitFuncDecl(swift::FuncDecl*) + 148
11 swift-frontend           0x0000000102f9cf5c (anonymous namespace)::DeclChecker::visit(swift::Decl*) + 224
12 swift-frontend           0x0000000102f9ce68 swift::TypeChecker::typeCheckDecl(swift::Decl*, bool) + 116
13 swift-frontend           0x0000000103049d00 swift::TypeCheckSourceFileRequest::evaluate(swift::Evaluator&, swift::SourceFile*) const + 176
14 swift-frontend           0x000000010304b9e4 llvm::Expected<swift::TypeCheckSourceFileRequest::OutputType> swift::Evaluator::getResultUncached<swift::TypeCheckSourceFileRequest>(swift::TypeCheckSourceFileRequest const&) + 384
15 swift-frontend           0x000000010304b7b0 llvm::Expected<swift::TypeCheckSourceFileRequest::OutputType> swift::Evaluator::getResultCached<swift::TypeCheckSourceFileRequest, (void*)0>(swift::TypeCheckSourceFileRequest const&) + 108
16 swift-frontend           0x0000000103049b60 swift::TypeCheckSourceFileRequest::OutputType swift::evaluateOrDefault<swift::TypeCheckSourceFileRequest>(swift::Evaluator&, swift::TypeCheckSourceFileRequest, swift::TypeCheckSourceFileRequest::OutputType) + 28
17 swift-frontend           0x0000000102393274 bool llvm::function_ref<bool (swift::SourceFile&)>::callback_fn<swift::CompilerInstance::performSema()::$_6>(long, swift::SourceFile&) + 16
18 swift-frontend           0x000000010238d458 swift::CompilerInstance::forEachFileToTypeCheck(llvm::function_ref<bool (swift::SourceFile&)>) + 160
19 swift-frontend           0x000000010238d38c swift::CompilerInstance::performSema() + 76
20 swift-frontend           0x000000010233dd74 withSemanticAnalysis(swift::CompilerInstance&, swift::FrontendObserver*, llvm::function_ref<bool (swift::CompilerInstance&)>) + 56
21 swift-frontend           0x0000000102334930 swift::performFrontend(llvm::ArrayRef<char const*>, char const*, void*, swift::FrontendObserver*) + 3012
22 swift-frontend           0x000000010226e734 swift::mainEntry(int, char const**) + 484
23 dyld                     0x0000000118dc10f4 start + 520
zsh: abort      bin/swift ../../../hello_world_test/file.swift
```

AnthonyLatsis commented:
Super weird. Printing to stderr works, right? Did you somehow modify the build-script command from the guide when building from scratch? I'd rather not start thinking that this is somehow related to the OS/Xcode version or something along those lines.

philipturner commented:
> Did you somehow modify the build-script command from the guide when building from scratch?

I have no idea. That's probably the case, but hard resetting my local repository would consume more time than it's worth. Are you willing to help me investigate another minor crash with AutoDiff?

---

I'm investigating the crash at the link below:

https://github.com/philipturner/swift-differentiable-collection-experiments/blob/main/compiler_crashers/crash_5.swift

I have found out the following so far:
- If you remove the `@differentiable` attribute, it doesn't crash.
- If you swap the `Void` closure parameter for something else that isn't differentiable, e.g. `Int`, it doesn't crash.
- If you swap the `Void` for `()`, it crashes.
- It crashes during the SILGen stage. Looking at the AST, I don't see anything alarming. But, I may not know what I'm doing.

The files referenced at the end of the stack trace:
- /lib/sema/TypeCheckAttr.cpp
- /lib/SIL/IR/SILFunctionType.cpp
- /lib/SIL/IR/TypeLowering.cpp

Maybe we should move this conversation into another thread - the one we originated from?

AnthonyLatsis commented:
Can you try running the following command and see it the fatal error appears?

```
/Users/philipturner/Documents/Swift-Compiler/swift-project/build/Ninja-RelWithDebInfoAssert/swift-macosx-arm64/bin/swift-frontend -frontend -interpret ../../../hello_world_test/file.swift -Xllvm -aarch64-use-tbi -enable-objc-interop -sdk /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX12.1.sdk -color-diagnostics -resource-dir /Users/philipturner/Documents/Swift-Compiler/swift-project/build/Ninja-RelWithDebInfoAssert/swift-macosx-arm64/lib/swift -module-name file -target-sdk-version 12.1
```

philipturner commented:
I ran `ninja swift-frontend`, then your command, and no difference.

philipturner commented:
About the new bug I'm facing, I've narrowed it down further. The problem is in the AST stage. When I set the closure's parameter type to `Int`, the AST says `<<error type>>` there. But for `Void`, it doesn't. `Void` is just as incompatible with differentiation as `Int`, so shouldn't it be `<<error type>>` too?

It seems that it's skipping over something in /lib/Sema/TypeChecker.cpp, line 3027. It should make a call to `diagnoseInvalidFunctionType`, but it doesn't for `Void`.

AnthonyLatsis commented:
> About the new bug I'm facing, I've narrowed it down further. The problem is in the AST stage. When I set the closure's parameter type to `Int`, the AST says `<<error type>>` there. But for `Void`, it doesn't. `Void` is just as incompatible with differentiation as `Int`, so shouldn't it be `<<error type>>` too?

I'll have a look.

___

Out of curiosity, could you try placing the `report_fatal_error` at an earlier compilation stage, at the start of, say, `Parser::parseDeclFunc`, and running it against just a trivial function declaration like `func foo() {}`?

philipturner commented:
I'm currently figuring out what's happening at the place it's supposed to throw an error diagnostic. I found a very handy trick that makes debugging so much easier (and doesn't corrupt stdlib builds):
```
llvm::outs() << "HELLO WORLD!\n";
```
For example, I found out that it reaches line 580 in both cases (with `Int` and with `Void`). But, they should both also reach line 639 and they don't (TypeChecker.cpp).

AnthonyLatsis commented:
> Maybe we should move this conversation into another thread - the one we originated from?

You could file a bug so that we can move the discussion to JIRA.

CodaFi commented on lib/Sema/TypeCheckAttr.cpp:
```diff
@@ -4749,6 +4749,8 @@ static bool typeCheckDerivativeAttr(ASTContext &Ctx, Decl *D,
   // to be enabled.
   if (checkIfDifferentiableProgrammingEnabled(Ctx, attr, D->getDeclContext()))
     return true;
+  if (D->isInvalid())
```
Checking the invalid bit and skipping work in Sema is certainly *a way* to make crashes go away. But we probably want to identify a root cause and fix that first. There's very few instances where this pattern actually helps. In practice it can hinder many more processes downstream as it either hides bugs, shifts data dependencies, or hides dependencies from the incremental build. We're trying to eliminate usages of this pattern where we can - unfortunately many of them are load bearing.

AnthonyLatsis replied:
The root cause is that we're calling `setInvalid()` when a generic parameter is not used in the type signature of a function/subscript etc.

philipturner replied:
So the real question is: why was `typeCheckDerivativeAttr` called when an error diagnostic was issued?

Is this correct?

CodaFi replied:
> why was typeCheckDerivativeAttr called when an error diagnostic was issued?

We try not to skip performing semantic analysis just because the AST was malformed. The purpose of Sema is, after all, to explain to the user what went wrong. 

> The root cause is that we're calling setInvalid() when a generic parameter is not used in the type signature of a function/subscript etc., although there seems to be nothing wrong with leaving the interface type as-is.

I agree that this seems wrong. You could probably start by just dropping the call to `setInvalid()` on the path and seeing what changes.

CodaFi replied:
> The root cause is that we're calling setInvalid() when a generic parameter is not used in the type signature of a function/subscript etc., although there seems to be nothing wrong with leaving the interface type as-is.

Agreed. Start by eliminating that call and seeing what changes.

philipturner replied:
@AnthonyLatsis I have my mind focused on the new bug I'm to file on JIRA. Would you be able to investigate that call and send over the necessary changes via a code review?

AnthonyLatsis replied:
> Start by eliminating that call and seeing what changes.

Pretty reassuring, running tests locally shows only 2 new `generic parameter 'T' could not be inferred` errors that are now expected. 
___

@philipturner if you are willing to see this through, you have to delete [this](https://github.com/apple/swift/blob/cb27114b8853ace172e325fe47a086873b215367/lib/Sema/TypeCheckGeneric.cpp#L509) call and update the following failing tests (and add the AutoDiff regression test):

```
Swift(macosx-x86_64) :: Constraints/overload.swift
Swift(macosx-x86_64) :: Generics/unbound.swift
```

philipturner replied:
Got it! Thanks for narrowing down the crux of the bug!

philipturner replied:
> (and add the AutoDiff regression test):

That just means add a new test proving this bug was fixed, correct? If so, I'll need a bit of help setting that up.

AnthonyLatsis replied:
We already have a test file for `@derivative` checking: `../test/AutoDiff/Sema/derivative_attr_type_checking.swift`. Just add your test case to that and run `lit` on it to see the diagnostics that are expected. If the expectations are correct, incorporate the missing diagnostics into your test using `expected-*` directives, as shown elsewhere in that file. The lit command is

```
cd ../build/Ninja-RelWithDebInfoAssert/test-macosx-$(uname -m) && ../llvm-project/llvm/utils/lit/lit.py -sv --filter=derivative_attr_type_checking AutoDiff/Sema
```

philipturner replied:
@AnthonyLatsis done!

philipturner commented:
SR-15808

philipturner commented:
@AnthonyLatsis do you have to capacity to pick up where I left off on the JIRA thread? I have a hypothetical way to test the bug's true source. But, I don't have the experience to pull it off efficiently just yet.

philipturner commented:
I got the correct diagnostic using @AnthonyLatsis's suggested patch! It was hidden away by the bug!
```
../../../hello_world_test/file.swift:6:17: error: cannot find 'remainder' in scope
@derivative(of: remainder(_:_:))
```

philipturner commented:
@CodaFi btw racy_async_let_fibonacci just failed while I was running tests using this command:
```
# Run all tests under test/.
../llvm-project/llvm/utils/lit/lit.py -s -vv \
  ../build/Ninja-RelWithDebInfoAssert/swift-macosx-$(uname -m)/test-macosx-$(uname -m)
```

philipturner committed:
Fix bug and generics tests

AnthonyLatsis commented:
> I got the correct diagnostic using @AnthonyLatsis's suggested patch!

Well, if it wasn't for @CodaFi, I probably would not mind the root cause, so thanks to Robert!

> racy_async_let_fibonacci just failed

Should be unrelated.

philipturner commented:
> Should be unrelated.

Sorry. I had a thread about this with @CodaFi and someone else: #40875. We were trying to reproduce the flaky test failure.

philipturner committed:
Add AutoDiff regression test

AnthonyLatsis commented:
@swift-ci please smoke test macOS

philipturner commented:
This is my very first time fixing an AutoDiff bug!

philipturner commented:
macOS smoke test failed.

AnthonyLatsis commented:
@swift-ci please clean smoke test macOS

rxwei reviewed (commented):
With the latest changes, it is no longer a fix directly related to autodiff.  I'd suggest removing "[AutoDiff] " and modifying the PR description to describe how this affects non-autodiff Swift diagnostics.

philipturner commented:
Is that sufficient?

rxwei commented:
> modifying the PR description to describe how this affects non-autodiff Swift diagnostics.

philipturner commented:
Fixed. Also, please tell me whether I have pinged you too much over the last week. It's affecting how well I can collaborate with you, probably making it more hard on you than anything.

If I am, I'd like to know so that I can tone down the number of pull requests. Or, if it's fine, then I can CC most of my pull requests to you.

CodaFi reviewed (approved):
Looks much better.

CodaFi commented:
@swift-ci clean smoke test

rxwei commented:
Could you please squash your changes into a single commit?

AnthonyLatsis commented:
We can squash and merge to avoid rerunning tests.
